hydra-node/src/Hydra/Network/Reliability.hs

```diff
@@ -133,19 +134,28 @@ instance Arbitrary ReliabilityLog where
 -- NOTE: Make better use of Vectors? We should perhaps use a `MVector` to be able to
 -- mutate in-place and not need `zipWith`
 withReliability ::
-  (MonadThrow (STM m), MonadThrow m, MonadAsync m) =>
+  (MonadThrow (STM m), MonadThrow m, MonadAsync m, ToJSON msg, FromJSON msg) =>
   -- | Tracer for logging messages.
   Tracer m ReliabilityLog ->
+  -- | Persistence handle to store messages
+  PersistenceIncremental (Heartbeat msg) m ->
+  -- | Persistence handle to store acks
+  Persistence (Vector Int) m ->
   -- | Our own party identifier.
   Party ->
   -- | Other parties' identifiers.
   [Party] ->
   -- | Underlying network component providing consuming and sending channels.
   NetworkComponent m (Authenticated (ReliableMsg (Heartbeat msg))) (ReliableMsg (Heartbeat msg)) a ->
   NetworkComponent m (Authenticated (Heartbeat msg)) (Heartbeat msg) a
-withReliability tracer me otherParties withRawNetwork callback action = do
-  ackCounter <- newTVarIO $ replicate (length allParties) 0
-  sentMessages <- newTVarIO IMap.empty
+withReliability tracer msgPersistence ackPersistence me otherParties withRawNetwork callback action = do
+  startingAckCounter <-
+    load ackPersistence >>= \case
+      Nothing -> pure $ replicate (length allParties) 0
+      Just existingCounter -> pure existingCounter
+  ackCounter <- newTVarIO startingAckCounter
+  storedMessages <- loadAll msgPersistence
+  sentMessages <- newTVarIO $ IMap.fromList (zip [1 ..] storedMessages)
```

Comment:
If we store the messages persistently then we don't need to store them in memory. We should rather read them back from persistent storage in the `resend` handler.